File: worker/src/worker/storage.py
"""Storage utilities for persisting results back to Redis."""
from __future__ import annotations

import json
import logging
from datetime import datetime, timezone
from typing import Any, Dict, List

from .config import get_settings
from .logger import get_logger, log_with_context
from .metrics import worker_chunks_failed_total, worker_io_time_seconds
from .redis_client import RedisClient
from .types import ChunkResult, ClusterResult, FailureRecord
from .utils import timer

logger = get_logger(__name__)


class ResultStorage:
    """Store chunk results and failures in Redis with instrumentation."""

    def __init__(self, redis_client: RedisClient, worker_id: str) -> None:
        self._redis = redis_client
        self._settings = get_settings()
        self._worker_id = worker_id

    async def push_result(self, brand: str, result: ChunkResult) -> float:
        key = f"{self._settings.redis_result_prefix}:{brand}:chunks"
        payload = json.dumps(self._format_for_orchestrator(result))
        with timer() as timing:
            await self._redis.rpush(key, payload)
        elapsed_ms = timing["elapsed_ms"]
        worker_io_time_seconds.labels(self._worker_id, brand, "push").observe(elapsed_ms / 1000)
        log_with_context(
            logger,
            level=logging.INFO,
            message="Result pushed to Redis",
            context={
                "worker_id": self._worker_id,
                "brand": brand,
                "key": key,
                "chunk_id": result.chunk_id,
            },
            metrics={"push_time_ms": elapsed_ms},
        )
        return elapsed_ms

    def _format_for_orchestrator(self, result: ChunkResult) -> Dict[str, Any]:
        clusters_payload = self._build_clusters(result.clusters)
        sentiment = self._aggregate_sentiment(result.clusters)
        topics = self._extract_topics(result.clusters)
        spike_detected = any(cluster.get("spike", False) for cluster in clusters_payload)
        mention_count = sum(cluster["mentionCount"] for cluster in clusters_payload)

        return {
            "chunkId": result.chunk_id,
            "brand": result.brand,
            "processedAt": datetime.now(timezone.utc).isoformat(),
            "sentiment": sentiment,
            "clusters": clusters_payload,
            "topics": topics,
            "summary": self._combine_summaries(result.clusters),
            "spikeDetected": spike_detected,
            "meta": {
                "metrics": result.metrics.model_dump(),
                "mentionCount": mention_count,
            },
        }

    def _build_clusters(self, clusters: List[ClusterResult]) -> List[Dict[str, Any]]:
        cluster_payload: List[Dict[str, Any]] = []
        for cluster in clusters:
            sentiment = cluster.sentiment or {}
            sentiment_score = float(sentiment.get("positive", 0.0)) - float(sentiment.get("negative", 0.0))
            label = self._normalize_summary_text(cluster.summary, cluster.examples, fallback_label=f"Cluster {cluster.cluster_id}")
            cluster_payload.append(
                {
                    "id": str(cluster.cluster_id),
                    "label": label,
                    "mentions": cluster.examples,
                    "sentimentScore": sentiment_score,
                    "spike": cluster.spike,
                    "mentionCount": cluster.count,
                }
            )
        return cluster_payload

    def _aggregate_sentiment(self, clusters: List[ClusterResult]) -> Dict[str, float]:
        totals = {"positive": 0.0, "neutral": 0.0, "negative": 0.0}
        counted = 0
        for cluster in clusters:
            if not cluster.sentiment:
                continue
            counted += 1
            totals["positive"] += float(cluster.sentiment.get("positive", 0.0))
            totals["neutral"] += float(cluster.sentiment.get("neutral", 0.0))
            totals["negative"] += float(cluster.sentiment.get("negative", 0.0))

        if counted > 0:
            for key in totals:
                totals[key] /= counted

        totals["score"] = totals["positive"] - totals["negative"]
        return totals

    def _extract_topics(self, clusters: List[ClusterResult]) -> List[str]:
        topics: List[str] = []
        for cluster in clusters:
            normalized = self._normalize_summary_text(cluster.summary, cluster.examples)
            if normalized:
                topics.append(normalized)
            elif cluster.examples:
                topics.extend(cluster.examples[:1])
        return [topic.strip() for topic in topics if topic.strip()][:10]

    def _combine_summaries(self, clusters: List[ClusterResult]) -> str:
        lines: List[str] = []
        for cluster in clusters:
            normalized = self._normalize_summary_text(cluster.summary, cluster.examples)
            if normalized:
                lines.append(normalized)
        if not lines:
            return ""
        return " ".join(lines)

    def _normalize_summary_text(self, summary: str | None, examples: List[str], *, fallback_label: str | None = None) -> str:
        candidate = (summary or "").strip()
        if candidate.startswith("{") and candidate.endswith("}") and "positive" in candidate and "negative" in candidate:
            candidate = ""
        if not candidate and examples:
            candidate = examples[0].strip()
        if not candidate and fallback_label:
            candidate = fallback_label
        return candidate

    async def record_failure(self, brand: str, failure: FailureRecord, *, reason_label: str) -> float:
        key = f"{self._settings.redis_failed_prefix}:{brand}"
        payload = failure.model_dump_json()
        with timer() as timing:
            await self._redis.record_failure(key, payload)
        elapsed_ms = timing["elapsed_ms"]
        worker_chunks_failed_total.labels(self._worker_id, brand, reason_label).inc()
        worker_io_time_seconds.labels(self._worker_id, brand, "failure").observe(elapsed_ms / 1000)
        log_with_context(
            logger,
            level=logging.WARNING,
            message="Failure recorded",
            context={
                "worker_id": self._worker_id,
                "brand": brand,
                "chunk_id": failure.chunk_id,
                "reason": failure.reason,
            },
            metrics={"failure_record_time_ms": elapsed_ms},
        )
        return elapsed_ms

File: worker-rs/src/storage.rs
use std::sync::Arc;
use std::time::Instant;

use anyhow::Context;
use chrono::Utc;
use serde_json::json;
use tracing::info;

use crate::config::Settings;
use crate::metrics::{
    WORKER_CHUNKS_FAILED_TOTAL, WORKER_CHUNKS_PROCESSED_TOTAL, WORKER_IO_TIME_SECONDS,
};
use crate::redis_client::RedisClient;
use crate::types::{ChunkResult, FailureRecord};

pub struct ResultStorage {
    redis: RedisClient,
    settings: Arc<Settings>,
}

impl ResultStorage {
    pub fn new(redis: RedisClient, settings: Arc<Settings>) -> Self {
        Self { redis, settings }
    }

    pub async fn push_result(&self, brand: &str, result: &mut ChunkResult) -> anyhow::Result<f64> {
        let key = format!("{}:{}:chunks", self.settings.redis_result_prefix, brand);
        let payload = self.format_for_orchestrator(result);
        let payload_str = serde_json::to_string(&payload).context("serialise chunk result")?;

        let start = Instant::now();
        self.redis.rpush(&key, &payload_str).await?;
        let elapsed_ms = start.elapsed().as_secs_f64() * 1000.0;

        result.metrics.io_time_ms += elapsed_ms;

        WORKER_IO_TIME_SECONDS
            .with_label_values(&[&self.settings.worker_id, brand, "push"])
            .observe(elapsed_ms / 1000.0);
        WORKER_CHUNKS_PROCESSED_TOTAL
            .with_label_values(&[&self.settings.worker_id, brand])
            .inc();

        info!(
            worker_id = %self.settings.worker_id,
            brand, key, chunk_id = %result.chunk_id,
            push_time_ms = elapsed_ms,
            "Result pushed to Redis"
        );

        Ok(elapsed_ms)
    }

    pub async fn record_failure(
        &self,
        brand: &str,
        failure: &FailureRecord,
        reason_label: &str,
    ) -> anyhow::Result<f64> {
        let key = format!("{}:{}", self.settings.redis_failed_prefix, brand);
        let payload = serde_json::to_string(failure).context("serialise failure record")?;

        let start = Instant::now();
        self.redis.record_failure(&key, &payload).await?;
        let elapsed_ms = start.elapsed().as_secs_f64() * 1000.0;

        WORKER_CHUNKS_FAILED_TOTAL
            .with_label_values(&[&self.settings.worker_id, brand, reason_label])
            .inc();
        WORKER_IO_TIME_SECONDS
            .with_label_values(&[&self.settings.worker_id, brand, "failure"])
            .observe(elapsed_ms / 1000.0);

        info!(
            worker_id = %self.settings.worker_id,
            brand,
            chunk_id = %failure.chunk_id,
            reason = %failure.reason,
            failure_record_time_ms = elapsed_ms,
            "Failure recorded",
        );

        Ok(elapsed_ms)
    }

    fn format_for_orchestrator(&self, result: &ChunkResult) -> serde_json::Value {
        let sentiment = self.aggregate_sentiment(&result.clusters);
        let topics = self.extract_topics(&result.clusters);
        let spike_detected = result.clusters.iter().any(|cluster| cluster.spike);
        let mention_count: usize = result.clusters.iter().map(|cluster| cluster.count).sum();

        json!({
            "chunkId": result.chunk_id,
            "brand": result.brand,
            "processedAt": Utc::now().to_rfc3339(),
            "sentiment": sentiment,
            "clusters": self.build_clusters(&result.clusters),
            "topics": topics,
            "summary": self.combine_summaries(&result.clusters),
            "spikeDetected": spike_detected,
            "meta": {
                "metrics": result.metrics,
                "mentionCount": mention_count,
            }
        })
    }

    fn build_clusters(&self, clusters: &[crate::types::ClusterResult]) -> Vec<serde_json::Value> {
        clusters
            .iter()
            .map(|cluster| {
                let sentiment_score = cluster
                    .sentiment
                    .get("positive")
                    .copied()
                    .unwrap_or_default()
                    - cluster.sentiment.get("negative").copied().unwrap_or_default();
                let label = self.normalise_summary_text(
                    cluster.summary.as_deref(),
                    &cluster.examples,
                    Some(format!("Cluster {}", cluster.cluster_id)),
                );
                json!({
                    "id": cluster.cluster_id.to_string(),
                    "label": label,
                    "mentions": cluster.examples,
                    "sentimentScore": sentiment_score,
                    "spike": cluster.spike,
                    "mentionCount": cluster.count,
                })
            })
            .collect()
    }

    fn aggregate_sentiment(
        &self,
        clusters: &[crate::types::ClusterResult],
    ) -> serde_json::Value {
        let mut totals = [0.0f32, 0.0, 0.0]; // positive, neutral, negative
        let mut counted = 0.0f32;
        for cluster in clusters {
            if !cluster.sentiment.is_empty() {
                totals[0] += cluster.sentiment.get("positive").copied().unwrap_or_default();
                totals[1] += cluster.sentiment.get("neutral").copied().unwrap_or_default();
                totals[2] += cluster.sentiment.get("negative").copied().unwrap_or_default();
                counted += 1.0;
            }
        }
        if counted > 0.0 {
            totals.iter_mut().for_each(|value| *value /= counted);
        }
        let score = totals[0] - totals[2];
        json!({
            "positive": totals[0],
            "neutral": totals[1],
            "negative": totals[2],
            "score": score,
        })
    }

    fn extract_topics(&self, clusters: &[crate::types::ClusterResult]) -> Vec<String> {
        let mut topics: Vec<String> = Vec::new();
        for cluster in clusters {
            if let Some(normalised) = self
                .normalise_summary_text(cluster.summary.as_deref(), &cluster.examples, None)
                .filter(|value| !value.is_empty())
            {
                topics.push(normalised);
            } else if let Some(example) = cluster.examples.get(0) {
                topics.push(example.clone());
            }
        }
        topics.truncate(10);
        topics
    }

    fn combine_summaries(&self, clusters: &[crate::types::ClusterResult]) -> String {
        clusters
            .iter()
            .filter_map(|cluster| {
                self.normalise_summary_text(cluster.summary.as_deref(), &cluster.examples, None)
            })
            .filter(|text| !text.is_empty())
            .collect::<Vec<_>>()
            .join(" ")
    }

    fn normalise_summary_text(
        &self,
        summary: Option<&str>,
        examples: &[String],
        fallback_label: Option<String>,
    ) -> Option<String> {
        let mut candidate = summary.unwrap_or_default().trim().to_string();
        if candidate.starts_with('{') && candidate.ends_with('}') && candidate.contains("positive") {
            candidate.clear();
        }
        if candidate.is_empty() {
            if let Some(example) = examples.get(0) {
                candidate = example.trim().to_string();
            }
        }
        if candidate.is_empty() {
            if let Some(fallback) = fallback_label {
                candidate = fallback;
            }
        }
        Some(candidate)
    }
}
